include/OswLogger.h

```diff
@@ -0,0 +1,120 @@
+        this->log(file, line, severity_t::W, std::forward<T>(message)...);
+    };
+
+    template<typename... T>
+    inline void info(const char* file, const unsigned int line, T&&... message) {
+        this->log(file, line, severity_t::I, std::forward<T>(message)...);
+    };
+
+    template<typename... T>
+    inline void debug(const char* file, const unsigned int line, T&&... message) {
+        this->log(file, line, severity_t::D, std::forward<T>(message)...);
+    };
+  private:
+    static OswLogger* instance;
+    std::mutex m_lock;
+    enum class severity_t { D, I, W, E };
+
+    // Used to process the variadic arguments in a type-safe (and aware) manner -> https://stackoverflow.com/a/15711171
+    inline void do_in_order() {};
+    template<typename Lambda0, typename... Lambdas>
+    inline void do_in_order( Lambda0&& L0, Lambdas&&... Ls ) {
+        std::forward<Lambda0>(L0)();
+        do_in_order( std::forward<Lambdas>(Ls)... );
+    };
+
+    template<typename... T>
+    void log(const char* file, const unsigned int line, const severity_t severity, T&&... message) {
+        std::lock_guard<std::mutex> guard(this->m_lock);
+        this->prefix(file, line, severity);
+
+        do_in_order([&]() {
+            // If performance is no issue, we could just use String(...) and treat everything the same for the '\n'-iteration...
+            if constexpr (std::is_same<T, String>::value or std::is_same<T, std::string>::value or
+                std::is_same<T, String&>::value or std::is_same<T, std::string&>::value) {
+                // Iterate over message to find '\n', which trigger new lines...
+                for(auto& c : message) {
+                    if (c == '\n') {
+                        Serial.println();
+                        this->prefix(file, line, severity);
+                    } else
+                        Serial.print(c);
+                }
+            } else if constexpr (std::is_same<T, const char*>::value or std::is_same<T, char*>::value) {
+                // Iterate over message to find '\n', which trigger new lines...
+                for(auto& c : std::string_view(message)) {
+                    if (c == '\n') {
+                        Serial.println();
+                        this->prefix(file, line, severity);
+                    } else
+                        Serial.print(c);
+                }
+            } else
+                Serial.print(message);
+        }...);
+    
+        Serial.println();
+    };
+
+    void prefix(const char* file, const unsigned int line, const severity_t severity) {
+        if(severity == severity_t::D)
+            Serial.print("D: ");
+        else if(severity == severity_t::I)
+            Serial.print("I: ");
+        else if(severity == severity_t::W)
+            Serial.print("W: ");
+        else if(severity == severity_t::E)
+            Serial.print("E: ");
+        else
+            throw std::logic_error("Unknown severity level");
```

Comment:
How about `switch-case`

Reply:
In Kotlin we could talk about that, but in C++? No... This is one of these classic topics :roll_eyes: . I think switch-case is only useful if you want to use its fall-trough feature, in any other case the danger of missing only one `break` is too high - causing unpredictable behavior. Ofc we could discuss, that I even omitted the `{}`, but that will be fixed by the auto-formatter anyways...

**BUT!** I've just read this article: https://www.scaler.com/topics/c/difference-between-if-else-and-switch/
They mention the creation of jump-tables, which will improve performance :star_struck: 

Soooo... Proposal: Let this one pass for now and I'll create a PR down the road to migrate as-much-as-possible to `switch`-statements inside the whole OS.